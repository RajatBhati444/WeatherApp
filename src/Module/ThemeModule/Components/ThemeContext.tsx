import React, {createContext, useContext, useState, useCallback} from 'react';
import {THEME_TYPE} from '../Types/CommonTypes';
import {DarkTheme, DayTheme} from '../Config/Theme';

// Define the shape of the context
interface ThemeContextProps {
  theme: THEME_TYPE;
  setTheme: (theme: THEME_TYPE) => void;
  toggleTheme: () => void;
}

// Create the context
const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

// Create a provider
export const ThemeProvider: React.FC<{children: React.ReactNode}> = ({
  children,
}) => {
  const [theme, setTheme] = useState<THEME_TYPE>(DayTheme);

  const toggleTheme = useCallback(() => {
    setTheme(prevTheme => (prevTheme.type === 'dark' ? DayTheme : DarkTheme));
  }, []);

  const value = {theme, setTheme, toggleTheme};

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
};

// Custom hook to use the theme
export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
