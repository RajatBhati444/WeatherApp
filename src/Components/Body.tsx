import React, {useMemo} from 'react';
import {
  KeyboardAvoidingView,
  ScrollView,
  ScrollViewProps,
  StyleSheet,
} from 'react-native';
import {useTheme} from '../Module/ThemeModule/Components/ThemeContext';
import {VARIANT} from '../Module/ThemeModule/Types/CommonTypes';

interface BodyProps extends ScrollViewProps {
  backgroundColor?: string;
  children?: any;
  variant?: VARIANT;
}

function Body(props: BodyProps) {
  const {style, backgroundColor, variant} = props;
  const {theme} = useTheme();

  const styles = useMemo(
    () =>
      StyleSheet.create({
        containerStyle: {
          flexGrow: 1,
          backgroundColor: backgroundColor
            ? backgroundColor
            : variant
            ? theme.colors[variant]
            : theme.colors.background,
        },
      }),
    [backgroundColor, theme.colors, variant],
  );
  return (
    <KeyboardAvoidingView style={{flex: 1}}>
      <ScrollView
        contentContainerStyle={[styles.containerStyle, style]}
        keyboardShouldPersistTaps={'handled'}
        //@ts-ignore
        enableOnAndroid={false}
        showsVerticalScrollIndicator={false}
        {...props}
      />
    </KeyboardAvoidingView>
  );
}

export default Body;
