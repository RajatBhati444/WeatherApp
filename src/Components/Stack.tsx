import React, {forwardRef, memo, ReactNode, useMemo} from 'react';
import {Platform, StyleSheet, View, ViewProps, ViewStyle} from 'react-native';
import {VARIANT} from '../Module/ThemeModule/Types/CommonTypes';
import {useTheme} from '../Module/ThemeModule/Components/ThemeContext';

export type StackProps = {
  children?: ReactNode;
  variant?: VARIANT;
  padding?: number;
  paddingBottom?: number;
  paddingEnd?: number;
  paddingHorizontal?: number;
  paddingLeft?: number;
  paddingRight?: number;
  paddingStart?: number;
  paddingTop?: number;
  paddingVertical?: number;
  margin?: number;
  marginBottom?: number;
  marginEnd?: number;
  marginHorizontal?: number;
  marginLeft?: number;
  marginRight?: number;
  marginStart?: number;
  marginTop?: number;
  marginVertical?: number;
} & ViewProps &
  ViewStyle;

function Stack(props: StackProps, ref: React.LegacyRef<View>) {
  const {
    children,
    elevation = 0,
    style,
    variant,
    backgroundColor,
    onLayout,
    pointerEvents,
    testID,
    padding,
    paddingBottom,
    paddingEnd,
    paddingHorizontal,
    paddingLeft,
    paddingRight,
    paddingStart,
    paddingTop,
    paddingVertical,
    margin,
    marginBottom,
    marginEnd,
    marginHorizontal,
    marginLeft,
    marginRight,
    marginStart,
    marginTop,
    marginVertical,
    ...styleProps
  } = props;
  const {theme} = useTheme();

  const iosShadowElevation = useMemo(
    () =>
      elevation === 0
        ? {}
        : {
            shadowOpacity: 0.0015 * elevation + 0.18,
            shadowRadius: 0.54 * elevation,
            shadowOffset: {
              height: 0.6 * elevation,
              width: 0,
            },
            shadowColor: theme.colors.surface,
          },
    [elevation, theme.colors.surface],
  );

  const memoisedStyles = useMemo(
    () =>
      StyleSheet.create({
        viewStyles: {
          backgroundColor: backgroundColor
            ? backgroundColor
            : variant
            ? theme.colors[variant]
            : theme.colors.transparent,
          padding: padding,
          paddingBottom: paddingBottom,
          paddingEnd: paddingEnd,
          paddingHorizontal: paddingHorizontal,
          paddingLeft: paddingLeft,
          paddingRight: paddingRight,
          paddingStart: paddingStart,
          paddingTop: paddingTop,
          paddingVertical: paddingVertical,
          margin: margin,
          marginBottom: marginBottom,
          marginEnd: marginEnd,
          marginHorizontal: marginHorizontal,
          marginLeft: marginLeft,
          marginRight: marginRight,
          marginStart: marginStart,
          marginTop: marginTop,
          marginVertical: marginVertical,
        },
      }),
    [
      backgroundColor,
      margin,
      marginBottom,
      marginEnd,
      marginHorizontal,
      marginLeft,
      marginRight,
      marginStart,
      marginTop,
      marginVertical,
      padding,
      paddingBottom,
      paddingEnd,
      paddingHorizontal,
      paddingLeft,
      paddingRight,
      paddingStart,
      paddingTop,
      paddingVertical,
      theme.colors,
      variant,
    ],
  );

  return (
    <View
      ref={ref}
      testID={testID}
      onLayout={onLayout}
      pointerEvents={pointerEvents}
      style={[
        Platform.OS === 'ios' ? iosShadowElevation : {elevation},
        styleProps,
        memoisedStyles.viewStyles,
        style,
      ]}>
      {children}
    </View>
  );
}

export default memo(forwardRef(Stack));
