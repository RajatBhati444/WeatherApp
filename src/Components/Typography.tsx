import React, {forwardRef, memo, ReactNode} from 'react';
import {Text, TextProps, TextStyle} from 'react-native';
import {VARIANT} from '../Module/ThemeModule/Types/CommonTypes';
import {useTheme} from '../Module/ThemeModule/Components/ThemeContext';

export type TypographyProps = {
  variant?: VARIANT;
  children?: ReactNode;
} & TextProps &
  TextStyle;

function Typography(
  props: TypographyProps,
  ref: React.LegacyRef<Text> | undefined,
) {
  const {
    style,
    color,
    textDecorationColor = color,

    variant,
    children,
    testID,
    accessibilityHint,
    accessibilityLabel,
    accessibilityRole,
    accessibilityState,
    accessible,
    adjustsFontSizeToFit,
    allowFontScaling = false,
    ellipsizeMode,
    maxFontSizeMultiplier,
    numberOfLines,
    onLongPress,
    onLayout,
    selectionColor,
    onTextLayout,
    selectable,
    textBreakStrategy,
    onPress,
    ...textStyleProps
  } = props;
  const {theme} = useTheme();

  return (
    <Text
      ref={ref}
      onPress={onPress}
      testID={testID}
      accessibilityHint={accessibilityHint}
      accessibilityLabel={accessibilityLabel}
      accessibilityRole={accessibilityRole}
      accessibilityState={accessibilityState}
      accessible={accessible}
      adjustsFontSizeToFit={adjustsFontSizeToFit}
      allowFontScaling={allowFontScaling}
      ellipsizeMode={ellipsizeMode}
      maxFontSizeMultiplier={maxFontSizeMultiplier}
      numberOfLines={numberOfLines}
      onLongPress={onLongPress}
      onLayout={onLayout}
      selectionColor={selectionColor}
      onTextLayout={onTextLayout}
      selectable={selectable}
      textBreakStrategy={textBreakStrategy}
      style={[
        textStyleProps,
        {
          color: color
            ? color
            : variant
            ? theme.colors[variant]
            : theme.colors.text,
          textDecorationColor: textDecorationColor
            ? textDecorationColor
            : variant
            ? theme.colors[variant]
            : theme.colors.text,
        },
        style,
      ]}>
      {children}
    </Text>
  );
}

export default memo(forwardRef(Typography));
