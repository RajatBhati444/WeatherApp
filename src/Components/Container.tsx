import {useFocusEffect} from '@react-navigation/core';
import React, {Fragment, useCallback} from 'react';
import {Platform, StatusBar, StatusBarStyle} from 'react-native';

import Stack from './Stack';
import {VARIANT} from '../Module/ThemeModule/Types/CommonTypes';
import {useTheme} from '../Module/ThemeModule/Components/ThemeContext';
import SafeAreaStack from './SafeAreaStack';

export type ContainerProps = {
  children?: React.ReactNode;
  backgroundColor?: string;
  statusBarBackgroundColor?: string;
  statusBarStyle?: StatusBarStyle;
  fullScreen?: boolean;
  variant?: VARIANT;
};

function Container(props: ContainerProps) {
  const {
    children,
    backgroundColor,
    fullScreen = true,
    statusBarBackgroundColor = '#251404',
    statusBarStyle = 'light-content',
    variant,
  } = props;
  const {theme} = useTheme();
  const statusBarBackgroundColorIos =
    statusBarBackgroundColor ??
    (fullScreen ? theme.colors.transparent : theme.colors.transparent);
  const screenBackgroundColor = backgroundColor
    ? backgroundColor
    : variant
    ? theme.colors[variant]
    : theme.colors.surface;

  useFocusEffect(
    useCallback(() => {
      if (Platform.OS === 'android') {
        StatusBar.setTranslucent(fullScreen ?? false);
        StatusBar.setBackgroundColor(
          statusBarBackgroundColor ?? theme.colors.adaptivePrimary,
        );
      }
      StatusBar.setBarStyle(statusBarStyle);
    }, [
      fullScreen,
      statusBarBackgroundColor,
      statusBarStyle,
      theme.colors.adaptivePrimary,
    ]),
  );

  return (
    <Stack flex={1} backgroundColor={screenBackgroundColor}>
      {fullScreen ? (
        <Fragment>{children}</Fragment>
      ) : (
        <Fragment>
          <SafeAreaStack
            flex={0}
            backgroundColor={statusBarBackgroundColorIos}
          />
          <SafeAreaStack flex={1} backgroundColor={screenBackgroundColor}>
            {children}
          </SafeAreaStack>
        </Fragment>
      )}
    </Stack>
  );
}

export default Container;
