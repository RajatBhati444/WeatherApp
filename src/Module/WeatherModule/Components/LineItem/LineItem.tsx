import React from 'react';
import Stack from '../../../../Components/Stack';
import Typography from '../../../../Components/Typography';

export interface LineItemProps {
  title: string;
  value: string;
}

function LineItem({title, value}: LineItemProps) {
  return (
    <Stack
      alignItems="center"
      justifyContent="space-between"
      flexDirection="row">
      <Typography fontSize={16} fontWeight={'500'}>
        {title}
      </Typography>
      <Typography fontSize={13} fontWeight={'400'}>
        {value}
      </Typography>
    </Stack>
  );
}

export default React.memo(LineItem);
