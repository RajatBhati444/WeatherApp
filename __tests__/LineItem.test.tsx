import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import LineItem from '../src/Module/WeatherModule/Components/LineItem/LineItem';
import {ThemeProvider} from '../src/Module/ThemeModule/Components/ThemeContext';

test('LineItem renders correctly', async () => {
  await ReactTestRenderer.act(() => {
    const tree = ReactTestRenderer.create(
      <ThemeProvider>
        <LineItem title="Test title" value="4 kph" />
      </ThemeProvider>,
    );

    expect(tree).toMatchSnapshot();
  });
});
