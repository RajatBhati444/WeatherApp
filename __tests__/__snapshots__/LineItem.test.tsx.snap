// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineItem renders correctly 1`] = `null`;
